### PR TITLE
hledger 1.17.1.1

### PR DESCRIPTION
**Pull request by simonmichael**
This is an experimental stack-based formula for better reproducibility. See related discussion at
https://github.com/Homebrew/homebrew-core/pull/49010#issuecomment-574719702
https://github.com/Homebrew/brew/pull/6945
https://github.com/Homebrew/homebrew-core/pull/51183

### CODE REVIEW

**Comment by simonmichael** [comment]
Only the jenkins High Sierra build failed, possibly because of "18:55:48 install: mkdir /private/tmp/hledger-20200306-25397-1h25yay/hledger-1.17/.brew_home/.stack/programs/x86_64-osx/ghc-8.8.2/lib/ghc-8.8.2: File exists". Not sure why, could there be issues with isolation of filesystem between jobs ?

**Comment by Bo98** [comment]
The tmp directory name is randomised so I doubt it.

**Comment by simonmichael** [comment]
Re-pushed to see if it's repeatable.

**Review by chenrui333 — APPROVED** [review]
lgtm, love those comments.

File: Formula/hledger.rb
Comment (SMillerDev) [inline Formula/hledger.rb]: There is no way to tell stack to use a file? Because this bypasses all the checksum restrictions we have.

File: Formula/hledger.rb
Comment (SMillerDev) [inline Formula/hledger.rb]: This really needs better tests.

File: Formula/hledger.rb
Comment (SMillerDev) [inline Formula/hledger.rb]: If you want you can make this a PR comment. Please don't comment in the formula unless it's to describe a patch

File: Formula/hledger.rb
Comment (SMillerDev) [inline Formula/hledger.rb]: It should really use all the brew parts it can.

File: Formula/hledger.rb
Comment (SMillerDev) [inline Formula/hledger.rb]: Link these in the PR, not in comments.

File: Formula/hledger.rb
Comment (simonmichael) [reply Formula/hledger.rb]: Thanks for the review @SMillerDev. As you probably know this is an experimental formula exploring solutions for the problem discussed at https://github.com/Homebrew/brew/pull/6945.

Yes I believe I could make it use the checksummed tarballs for the hledger packages. All other haskell dependencies will still be fetched from stackage though. Homebrew's cabal-based formulae do this too, except those are less repeatable than fetching from stackage.

File: Formula/hledger.rb
Comment (simonmichael) [reply Formula/hledger.rb]: I agree. It's hard to test a curses app but I'm working on it and hope to improve this with some future release.

File: Formula/hledger.rb
Comment (simonmichael) [reply Formula/hledger.rb]: Ok, are you requiring all comments to be removed ? I ask for clarity, because this is a non-routine human-maintained formula, and these notes can help the next maintainer (who might no be me). Also I didn't see anything about it in the homebrew contrib docs.

File: Formula/hledger.rb
Comment (simonmichael) [reply Formula/hledger.rb]: Using brew's ghc will make it more inclined to break over time, but should be ok for now; I'll do that then.

**Comment by simonmichael** [comment]
The latest commit 2718526 is ready for review. It removes all but one comment, uses homebrew's ghc, uses the checksummed tarballs for the hledger packages, and passes strict audit.

File: Formula/hledger.rb
Comment (simonmichael) [reply Formula/hledger.rb]: Ok, can you explain - if I put them in the commit message, I'd be able to hunt them down with git blame - but do you mean in the PR body, or a PR comment ? If so, what's the process for a formula maintainer to find them ?

Comment (simonmichael) [reply Formula/hledger.rb]: I have moved them to both the commit message and the PR body. Also, I've added installation of man pages.

Comment (SMillerDev) [reply Formula/hledger.rb]: All merges indcate te PR that closed them. If it's in a comment it's easy to find. But commit boy will work too.

**Comment by simonmichael** [comment]
And one last push, installing the info manuals. This is ready for merge when CI passes.

**Comment by simonmichael** [comment]
All github CIs passed, but jenkins has been pending for over a day. I think it may have been confused by my force-pushing, because searching  in the Build History panel at https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests for this PR shows three past builds but none for the latest commit. Do you think we could ignore that and go ahead with merging this ?

**Review by chenrui333 — APPROVED** [review]
sounds like good to go?

File: Formula/hledger.rb
Comment (SMillerDev) [inline Formula/hledger.rb:1]: Is this still needed?

**Comment by SMillerDev** [comment]
There's just the test improvement to go.

File: Formula/hledger.rb
Comment (simonmichael) [reply Formula/hledger.rb]: You're right, it's not needed since I implemented [MikeMcQuaid's advice](https://github.com/Homebrew/brew/pull/6945#issuecomment-575311420). I've removed it.

**Comment by simonmichael** [comment]
@SMillerDev, re tests what do you have in mind ? Here's why I think it's ok not to block on this:

- as mentioned, hledger-ui is a curses-style app (and hledger-web is a web app). Testing the UI is hard (I've tried) and out of my scope/time budget for this release
- both apps are thin UI wrappers around the same core hledger code which is tested by "hledger test"
- there's some truth to the haskell "if it compiles, it works" motto; in my experience with these apps, if they run at all they will work fine
- homebrew users have been using this formula successfully for years

**Comment by SMillerDev** [comment]
> as mentioned, hledger-ui is a curses-style app (and hledger-web is a web app). Testing the UI is hard (I've tried) and out of my scope/time budget for this release

Web apps are pretty easy to test by running it and doing a request. See for example how we test requests to httpd. For the curses test you could see if it starts in a similar manner.

> both apps are thin UI wrappers around the same core hledger code which is tested by "hledger test"

That doesn't mean that the wrappers work though.

> there's some truth to the haskell "if it compiles, it works" motto; in my experience with these apps, if they run at all they will work fine

I'd rather not take that chance, especially with a major rewrite of the install.

> homebrew users have been using this formula successfully for years

And making proper tests ensures they can use it for years to come.

**Comment by simonmichael** [comment]
@SMillerDev I appreciate the thorough reviewing. Respectfully, if you insist on UI tests I think you're asking too much for a simple version bump in an existing homebrew package.

**Comment by SMillerDev** [comment]
You rewrote the entire install. That's hardly a simple version bump. I know this might might seem like a daunting task but it's really not that bad given the amount of prior art for web calls in homebrew test block.

**Comment by simonmichael** [comment]
@SMillerDev before I kick off another CI run, how does this test code look to you:
```
  test do
    system "#{bin}/hledger", "test"

    File.open(".hledger.journal", "w") {
      |f| f.write(<<~EOS
        2020/1/1
          boat  123
          cash
      EOS
    )}

    # assert_match /boat.+123/, shell_output("echo q | #{bin}/hledger-ui")
    system "#{bin}/hledger-ui", "--version"

    pid = fork do
      exec "hledger-web --serve"
    end
    sleep 1
    begin
      assert_match /boat +123/, shell_output("curl -s http://127.0.0.1:5000/journal")
    ensure
      Process.kill("SIGINT", pid)
      Process.wait(pid)
    end

  end
```
The hledger-ui test is unchanged as I don't know how to run an app that requires an interactive terminal in this script.

**Comment by SMillerDev** [comment]
Looks good, I'd probably do a run, sleep and kill for the ui app too.

**Comment by simonmichael** [comment]
I can't run hledger-ui from a script, it requires an interactive terminal:
```
Testing hledger
==> echo q | hledger-ui
hledger-ui: getTerminalAttributes: illegal operation (Inappropriate ioctl for device)
Killing child processes...
Error: hledger: failed
```

**Comment by simonmichael** [comment]
..so I've cleaned up per brew audit and pushed as is.

**Comment by SMillerDev** [comment]
Can't help that, thanks for trying. Tests look good now. 
Thanks @simonmichael! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Comment by simonmichael** [comment]
Thanks @SMillerDev, @chenrui333.

**Pull request closed without merging**
